server/src/main/java/org/elasticsearch/index/mapper/TimeSeriesRoutingIdFieldMapper.java

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.index.mapper;
+
+import org.apache.lucene.document.SortedDocValuesField;
+import org.apache.lucene.search.Query;
+import org.elasticsearch.common.util.ByteUtils;
+import org.elasticsearch.index.IndexMode;
+import org.elasticsearch.index.IndexVersions;
+import org.elasticsearch.index.fielddata.FieldData;
+import org.elasticsearch.index.fielddata.FieldDataContext;
+import org.elasticsearch.index.fielddata.IndexFieldData;
+import org.elasticsearch.index.fielddata.ScriptDocValues;
+import org.elasticsearch.index.fielddata.plain.SortedOrdinalsIndexFieldData;
+import org.elasticsearch.index.query.SearchExecutionContext;
+import org.elasticsearch.script.field.DelegateDocValuesField;
+import org.elasticsearch.search.aggregations.support.CoreValuesSourceType;
+
+import java.util.Base64;
+import java.util.Collections;
+
+/**
+ * Mapper for the {@code _seq_no} field.
+ *
+ * We expect to use the seq# for sorting, during collision checking and for
+ * doing range searches. Therefore the {@code _seq_no} field is stored both
+ * as a numeric doc value and as numeric indexed field.
+ *
+ * This mapper also manages the primary term field, which has no ES named
+ * equivalent. The primary term is only used during collision after receiving
+ * identical seq# values for two document copies. The primary term is stored as
+ * a doc value field without being indexed, since it is only intended for use
+ * as a key-value lookup.
+
+ */
+public class TimeSeriesRoutingIdFieldMapper extends MetadataFieldMapper {
+
+    public static final String NAME = "_routing_id";
```

Comment:
I think we should refer to routing id to routing hash instead? It isn't really an id and it is only used as a hash.
In that case the `TimeSeriesRoutingIdFieldMapper` should renamed to `TimeSeriesRoutingHashFieldMapper`, `_routing_id` should become `_routing_hash` and similar change in other places.